### ISSUE TITLE
Optional model imports collision resolver.

### ISSUE THREAD

**Issue by domandinho** [django-extensions/django-extensions#1116]
`shell_plus` imports all models on startup. 
But if two different models has exactly the same name, 
only one of them(this which was imported last) is accessible.
Both models should be accessible, one of them should be imported with alias.

For example suppose that we have two apps:
- languages with model Language (for example English)
- programming with model Language (for example Python)

I suggest to add
`MODEL_IMPORTS_COLLISION_RESOLVER` option to django settings, 
which is path to python function. 
This function will have one argument: `Dict[str, List[str]]`.

Key will be model name, value will be list of modules, which has model with this name.
Function will return `Dict[str, List[Union[str, Tuple[str, str]]]`, where key is module name, 
and value list of model names or tuples (model_name, alias). 
We use currently this structure to collect all model imports.

Two functions to resolve conflicts should be added:

- legacy function, which behaves exactly as now.
- app_name function which returns all imports from legacy version and aditionally aliased 
imports for models with duplicate name. It can also for performance reasons remove imports, 
which will be overwritten. 
It will concat app_name with model_name.

In previously defined example this function will receive:
```
{
	'Language': ['languages.models', 'programming.models'],
} + django models
```

and will return 
```
{
	'languages.models' [('Language', 'languages_Language')], 
	# import without alias from languages is not needed because it will be overwritten 
	'programming.models': ['Language', ('Language', 'programming_Language')]
	# import without alias from programming is needed, becase it won't be overwritten 
} + django models
```

We should also allow user to pass his own function, 
and describe in documentation what it should receive and return.

Is this feature is acceptable?
I want to implement this feature and submit PR if it is.

**Comment by trbs** [comment]
Does `SHELL_PLUS_MODEL_ALIASES` ( http://django-extensions.readthedocs.io/en/latest/shell_plus.html#configuration ) solve this problem as well ?

That said I think adding a user definable function to auto resolve conflicts is a great idea ! 

Lets also add one or two (example) function(s) in `django-extensions` as well, eg; auto resolving by prefixing app_name and/or resolve by preferring application to django or dependencies (like; `myapp.SomeModel` vs `thirdparty_app.SomeModel` which might be later in the module loading list). Which would be good for people to have something to play with and learn from and we can use it for the unit tests.

More then happy to accept PR for it and thanks for helping !

**Comment by domandinho** [comment]
Now we can try to resolve conflicts using:

`SHELL_PLUS_MODEL_ALIASES`:
Pros:
+ For single model developer can set any alias, which he want.

Cons:
- Developer need to find all conflicts manually and define aliases in configuration.
- Developer need to remember to add alias in settings, each time when creating model with conflicting name.
- This dictionary can be large if there are lot of collisions.

`SHELL_PLUS_APP_PREFIXES`:
Pros:
+ If prefixes for all apps are defined, new conflicts are resolved automatically.

Cons:
- Developer need to remember to add alias in settings, each time when creating new app or each time when first conflict in existing app occurced.
- When prefix is specified, ALL models from app are prefixed, and import without alias is disabled for them. 
So when i have 10 models in app and one of them has the conflicting name, than i need to access 9 other models using prefix.
- In most cases we need to add many similar items to this dictionary like: {<app_name>: <app_name>}
- User can define only prefix and _ separator is forced.

No solution is perfect so ability to add user defined function is needed. 
In my opinion it should has lowest priority and work only if:
- `SHELL_PLUS_MODEL_ALIASES` has no `<app_name>` key or `SHELL_PLUS_MODEL_ALIASES[<app_name>]` has no key <model_name>
- and `SHELL_PLUS_APP_PREFIXES` has no key `<app_name>`

**Comment by trbs** [comment]
I'm happy adding the a user definable function for this !

Another idea:

We can even call it: `MODEL_IMPORTS_RESOLVER`

And make it even more generic, being a callable which receives all the imports `shell_plus` itself found and want to do and returns said information. (Like you outlined above)

One function for this would be to resolve collisions on import errors, another would be to allow users to do all kinds of dynamic creation of the list of imports they want. (If they ignore the incoming list of imports they can override the entire behavior of imports of `shell_plus`)

User could also chain different functions, to do some dynamic resolving of imports then do collection fixing.

**Comment by domandinho** [comment]
PR https://github.com/django-extensions/django-extensions/pull/1117

**Comment by trbs** [comment]
Merged PR #1117
